pkg/apis/cartographer/v1alpha1/workload_source_printer.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2021 VMware, Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"io"
+	"strings"
+	"time"
+
+	metav1beta1 "k8s.io/apimachinery/pkg/apis/meta/v1beta1"
+
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/printer"
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/printer/table"
+)
+
+type resourcesQuery struct {
+	resourceName  string
+	outputName    string
+	imageHeader   string
+	updatedHeader string
+	rows          []metav1beta1.TableRow
+}
+
+func findResource(rq resourcesQuery, resources []RealizedResource) []metav1beta1.TableRow {
+	imageRow := metav1beta1.TableRow{}
+	updatedRow := metav1beta1.TableRow{}
+
+	for _, r := range resources {
+		for _, o := range r.Outputs {
+			if r.Name == rq.resourceName && o.Name == rq.outputName {
+				imageRow.Cells = append(imageRow.Cells,
+					rq.imageHeader,
+					strings.Replace(o.Preview, "\n", "", -1),
+				)
+				updatedRow.Cells = append(updatedRow.Cells,
+					rq.updatedHeader,
+					printer.TimestampSince(o.LastTransitionTime, time.Now()),
+				)
+				rq.rows = append(rq.rows, imageRow, updatedRow)
+				break
```

Comment:
This break is only exiting the inner loop, the outer loop will continue. It's fine to return here since there should only ever be a single match for a resource and output.

If there was ever a second match, the output from the earlier iteration would compound on itself. As `imageRow` and `updateRow` are defined outside of the loops and their cells are appended for each match, before the rows are added to the query's rows. For three matches, this would look something like:

```
image:    match1
update:   match1
image:    match1:   image:    match2
update:   match1:   update:   match2
image:    match1:   image:    match2:   image:    match3
update:   match1:   update:   match2:   update:   match3
```

It's best to define values as far inside the loop as possible, and never use values across multiple iterations of a loop when they are specified to a single iteration of that loop.